rust/agama-lib/src/storage/client/zfcp.rs

```diff
@@ -34,6 +34,60 @@ use crate::{
     },
 };
 
+pub async fn get_wwpns(controller_id: &str) -> Result<Vec<String>, ServiceError> {
+    let output = std::process::Command::new("zfcp_san_disc")
+        .arg("-b")
+        .arg(controller_id)
+        .arg("-W")
+        .output()
+        .expect("zfcp_san_disc failed"); // TODO: real error handling
+    Ok(String::from_utf8_lossy(&output.stdout)
+        .split_whitespace()
+        .map(|s| s.to_string())
+        .collect())
+}
+
+pub async fn get_luns(controller_id: String, wwpn: String) -> Result<Vec<String>, ServiceError> {
+    let output = std::process::Command::new("zfcp_san_disc")
+        .arg("-b")
+        .arg(controller_id)
+        .arg("-p")
+        .arg(wwpn)
+        .arg("-W")
+        .output()
+        .expect("zfcp_san_disc failed"); // TODO: real error handling
+    Ok(String::from_utf8_lossy(&output.stdout)
+        .split_whitespace()
+        .map(|s| s.to_string())
+        .collect())
+}
+
+/// Obtains a LUNs map for the given controller
+///
+/// Given a controller id it returns a HashMap with each of its WWPNs as keys and the list of
+/// LUNS corresponding to that specific WWPN as values.
+///
+/// Arguments:
+///
+/// `controller_id`: controller id
+pub async fn get_luns_map(
+    controller_id: &str,
+) -> Result<HashMap<String, Vec<String>>, ServiceError> {
+    let wwpns = get_wwpns(controller_id).await?;
+    let mut tasks = vec![];
+    for wwpn in wwpns {
+        tasks.push((
+            wwpn.clone(),
+            tokio::spawn(get_luns(controller_id.to_string(), wwpn.clone())),
```

Comment:
Instead of spawning all those processes, I would suggest giving [join_all](https://docs.rs/futures/latest/futures/future/fn.join_all.html) a try.

Reply:
Tokio has [JoinSet](https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html), which might be interested too.

Reply:
thanks for suggestion. I plan to have it as experiment first and this tokio spawn is suggested way in that linkedin rust training, so it is the first that I would like to give a try.

Reply:
and testing shows that my change somehow does not work, but as it is just learning/research project I will put it a bit on hold and focus on planned stuff and return to it probably next week.

Reply:
ok, I cannot sleep without debugging it...so now it is fixed and time is improved to 6.3 second..so already it goes down from 10.8 to 6.3 and I think it can be even better if I parallel also getting wwpns for controller. And then I can start playing with join_all and JoinSet